mockfirestore/transaction.py

```diff
@@ -0,0 +1,119 @@
+from functools import partial
+import random
+from typing import Iterable, Callable
+from string import ascii_letters, digits
+from .main import MockFirestore, DocumentReference, DocumentSnapshot, Query, Timestamp
+
+MAX_ATTEMPTS = 5
+ALPHABET = list(ascii_letters) + list(digits)
+_MISSING_ID_TEMPLATE = "The transaction has no transaction ID, so it cannot be {}."
+_CANT_BEGIN = "The transaction has already begun. Current transaction ID: {!r}."
+_CANT_ROLLBACK = _MISSING_ID_TEMPLATE.format("rolled back")
+_CANT_COMMIT = _MISSING_ID_TEMPLATE.format("committed")
+
+
+class MockWriteResult:
+    def __init__(self):
+        self.update_time = Timestamp.from_now()
+
+
+class Transaction:
+    """
+    This mostly follows the model from
+    https://googleapis.dev/python/firestore/latest/transaction.html
+    """
+    def __init__(self, client: MockFirestore,
+                 max_attempts=MAX_ATTEMPTS, read_only=False):
+        self._client = client
+        self._max_attempts = max_attempts
+        self._read_only = read_only
+        self._id = None
+        self._write_ops = []
+        self.write_results = None
+
+    @property
+    def in_progress(self):
+        return self._id is not None
+
+    @property
+    def id(self):
+        return self._id
+
+    def _begin(self, retry_id=None):
+        # generate a random ID to set the transaction as in_progress
+        self._id = ''.join(random.choices(ALPHABET, k=10))
+
+    def _clean_up(self):
+        self._write_ops.clear()
+        self._id = None
+
+    def _rollback(self):
+        if not self.in_progress:
+            raise ValueError(_CANT_ROLLBACK)
+
+        self._clean_up()
```

Comment:
This isn't called from anywhere, so unless you had another reason for including it I think it's best to leave it out until we've added some kind of mock implementation for rollbacks.

Reply:
This can get called by some parts of the real Firestore transaction machinery. So, if you monkeypatch the Transaction object to the mocked one, it is helpful to have all the methods so the tests still pass.

Reply:
This can get called by some parts of the real Firestore transaction machinery. So, if you monkeypatch the Transaction object to the mocked one, it is helpful to have all the methods so the tests still pass.

Reply:
That makes sense. It would be better to have it replicate an actual rollback in some way, but this is still much better than having no transaction support at all!

Reply:
That's a good idea. I added some thoughts on how we could implement rollbacks in https://github.com/mdowds/python-mock-firestore/issues/16, but don't have time right now to implement it. I might get to it at some point later.